Data race in the peertracker test suite

Our test suite is occasionally failing with a data race inside the peertracker test suite. It looks like the functions exposed by Testify aren't thread safe but we are calling them from inside a goroutine. Output below

```
==================
WARNING: DATA RACE
Read at 0x00c00009ca08 by goroutine 9:
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).TestAcceptDoesntFailWhenTrackerFails.func1()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:40 +0xb4
Previous write at 0x00c00009ca08 by goroutine 7:
  github.com/stretchr/testify/suite.(*Suite).SetT()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:35 +0x12f
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).SetT()
      <autogenerated>:1 +0x4d
  github.com/stretchr/testify/suite.Run.func2.1()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:124 +0xb9
  github.com/stretchr/testify/suite.Run.func2()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:127 +0x2ed
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
Goroutine 9 (running) created at:
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).TestAcceptDoesntFailWhenTrackerFails()
      /home/travis/gopath/src/github.com/spiffe/spire/pkg/common/peertracker/listener_test.go:84 +0x3c9
  runtime.call32()
      /home/travis/.build/.build-linux-x86_64/src/runtime/asm_amd64.s:519 +0x3a
  reflect.Value.Call()
      /home/travis/.build/.build-linux-x86_64/src/reflect/value.go:308 +0xc0
  github.com/stretchr/testify/suite.Run.func2()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:126 +0x2ec
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
Goroutine 7 (finished) created at:
  testing.(*T).Run()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:916 +0x65a
  github.com/stretchr/testify/suite.runTests()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:145 +0xe5
  github.com/stretchr/testify/suite.Run()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:132 +0x62b
  github.com/spiffe/spire/pkg/common/peertracker.TestListenerTestSuite()
      /home/travis/gopath/src/github.com/spiffe/spire/pkg/common/peertracker/listener_test.go:32 +0x5e
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
==================
==================
WARNING: DATA RACE
Read at 0x00c0001b6e80 by goroutine 9:
  github.com/stretchr/testify/require.(*Assertions).Error()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/require/require_forward.go:200 +0x3e
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).TestAcceptDoesntFailWhenTrackerFails.func1()
      /home/travis/gopath/src/github.com/spiffe/spire/pkg/common/peertracker/listener_test.go:86 +0x112
Previous write at 0x00c0001b6e80 by goroutine 7:
  github.com/stretchr/testify/suite.(*Suite).SetT()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/require/forward_requirements.go:11 +0xf7
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).SetT()
      <autogenerated>:1 +0x4d
  github.com/stretchr/testify/suite.Run.func2.1()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:124 +0xb9
  github.com/stretchr/testify/suite.Run.func2()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:127 +0x2ed
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
Goroutine 9 (running) created at:
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).TestAcceptDoesntFailWhenTrackerFails()
      /home/travis/gopath/src/github.com/spiffe/spire/pkg/common/peertracker/listener_test.go:84 +0x3c9
  runtime.call32()
      /home/travis/.build/.build-linux-x86_64/src/runtime/asm_amd64.s:519 +0x3a
  reflect.Value.Call()
      /home/travis/.build/.build-linux-x86_64/src/reflect/value.go:308 +0xc0
  github.com/stretchr/testify/suite.Run.func2()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:126 +0x2ec
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
Goroutine 7 (finished) created at:
  testing.(*T).Run()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:916 +0x65a
  github.com/stretchr/testify/suite.runTests()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:145 +0xe5
  github.com/stretchr/testify/suite.Run()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:132 +0x62b
  github.com/spiffe/spire/pkg/common/peertracker.TestListenerTestSuite()
      /home/travis/gopath/src/github.com/spiffe/spire/pkg/common/peertracker/listener_test.go:32 +0x5e
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
==================
==================
WARNING: DATA RACE
Read at 0x00c0001b6ec0 by goroutine 9:
  github.com/stretchr/testify/require.(*Assertions).Contains()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/require/require_forward.go:38 +0x3e
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).TestAcceptDoesntFailWhenTrackerFails.func1()
      /home/travis/gopath/src/github.com/spiffe/spire/pkg/common/peertracker/listener_test.go:87 +0x1c1
Previous write at 0x00c0001b6ec0 by goroutine 12:
  github.com/stretchr/testify/suite.(*Suite).SetT()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/require/forward_requirements.go:11 +0xf7
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).SetT()
      <autogenerated>:1 +0x4d
  github.com/stretchr/testify/suite.Run.func2()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:108 +0xc0
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
Goroutine 9 (running) created at:
  github.com/spiffe/spire/pkg/common/peertracker.(*ListenerTestSuite).TestAcceptDoesntFailWhenTrackerFails()
      /home/travis/gopath/src/github.com/spiffe/spire/pkg/common/peertracker/listener_test.go:84 +0x3c9
  runtime.call32()
      /home/travis/.build/.build-linux-x86_64/src/runtime/asm_amd64.s:519 +0x3a
  reflect.Value.Call()
      /home/travis/.build/.build-linux-x86_64/src/reflect/value.go:308 +0xc0
  github.com/stretchr/testify/suite.Run.func2()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:126 +0x2ec
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
Goroutine 12 (running) created at:
  testing.(*T).Run()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:916 +0x65a
  github.com/stretchr/testify/suite.runTests()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:145 +0xe5
  github.com/stretchr/testify/suite.Run()
      /home/travis/go/pkg/mod/github.com/stretchr/testify@v1.3.0/suite/suite.go:132 +0x62b
  github.com/spiffe/spire/pkg/common/peertracker.TestListenerTestSuite()
      /home/travis/gopath/src/github.com/spiffe/spire/pkg/common/peertracker/listener_test.go:32 +0x5e
  testing.tRunner()
      /home/travis/.build/.build-linux-x86_64/src/testing/testing.go:865 +0x163
==================
--- FAIL: TestListenerTestSuite (0.01s)
    --- FAIL: TestListenerTestSuite/TestAcceptFailsWhenUnderlyingAcceptFails (0.00s)
        testing.go:809: race detected during execution of test
    testing.go:809: race detected during execution of test
FAIL
FAIL	github.com/spiffe/spire/pkg/common/peertracker	1.900s
```

--- a/pkg/common/peertracker/listener_test.go
+++ b/pkg/common/peertracker/listener_test.go
@@ -1,6 +1,7 @@
 package peertracker
 
 import (
+	"context"
 	"errors"
 	"io/ioutil"
 	"net"
@@ -11,7 +12,6 @@ import (
 
 	"github.com/sirupsen/logrus"
 	"github.com/sirupsen/logrus/hooks/test"
-	"github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/suite"
 )
 
@@ -74,41 +74,69 @@ func (p *ListenerTestSuite) TestAcceptDoesntFailWhenTrackerFails() {
 	p.ul, err = lf.ListenUnix(p.unixAddr.Network(), p.unixAddr)
 	p.Require().NoError(err)
 
+	// used to cancel the log polling below if something goes wrong with
+	// the test
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
 	clientDone := make(chan struct{})
-	serverDone := make(chan struct{})
-	gotLog := make(chan struct{})
 	peer := newFakeUDSPeer(p.T())
 
 	peer.connect(p.unixAddr, clientDone)
 
+	type acceptResult struct {
+		conn net.Conn
+		err  error
+	}
+	acceptCh := make(chan acceptResult, 1)
 	go func() {
 		conn, err := p.ul.Accept()
-		p.Require().Error(err)
-		p.Require().Contains(err.Error(), "use of closed network connection")
-		p.Require().Nil(conn)
-		close(serverDone)
+		acceptCh <- acceptResult{
+			conn: conn,
+			err:  err,
+		}
 	}()
 
+	logCh := make(chan *logrus.Entry, 1)
 	go func() {
 		for {
 			logEntry := hook.LastEntry()
 			if logEntry == nil {
-				time.Sleep(time.Millisecond)
+				select {
+				case <-ctx.Done():
+					close(logCh)
+				case <-time.After(time.Millisecond * 10):
+				}
 				continue
 			}
-			p.Require().Equal("Connection failed during accept.", logEntry.Message)
-			logErr := logEntry.Data["error"]
-			p.Require().IsType(errors.New(""), logErr)
-			p.Require().EqualError(logErr.(error), "create new watcher failed")
-			close(gotLog)
-			break
+			logCh <- logEntry
 		}
 	}()
-	waitForChannelWithTimeout(p.Require(), gotLog, "waited too long for logs")
+
+	// Wait for the logs to show up demonstrating the accept failure
+	select {
+	case logEntry := <-logCh:
+		p.Require().NotNil(logEntry)
+		p.Require().Equal("Connection failed during accept.", logEntry.Message)
+		logErr := logEntry.Data["error"]
+		p.Require().IsType(errors.New(""), logErr)
+		p.Require().EqualError(logErr.(error), "create new watcher failed")
+	case <-time.After(time.Second):
+		p.Require().Fail("waited too long for logs")
+	}
 
 	p.Require().NoError(p.ul.Close())
 	p.ul = nil
-	waitForChannelWithTimeout(p.Require(), gotLog, "waited too long for server to close")
+
+	// Wait for the listener to stop
+	select {
+	case acceptRes := <-acceptCh:
+		p.Require().Error(acceptRes.err)
+		p.Require().Contains(acceptRes.err.Error(), "use of closed network connection")
+		p.Require().Nil(acceptRes.conn)
+	case <-time.After(time.Second):
+		p.Require().Fail("waited too long for listener to close")
+	}
 }
 
 func (p *ListenerTestSuite) TestAcceptFailsWhenUnderlyingAcceptFails() {
@@ -126,12 +154,4 @@ func (p *ListenerTestSuite) TestAcceptFailsWhenUnderlyingAcceptFails() {
 // returns an empty unix listener that will fail any call to Accept()
 func newFailingMockListenUnix(network string, laddr *net.UnixAddr) (*net.UnixListener, error) {
 	return &net.UnixListener{}, nil
-}
-
-func waitForChannelWithTimeout(require *require.Assertions, ch chan struct{}, failMsg string) {
-	select {
-	case <-ch:
-	case <-time.After(time.Second):
-		require.Fail(failMsg)
-	}
 }

